Document Kyma update process (#2216)
kyma-project/kyma: docs/kyma/docs/036-inst-update.md

---
title: Update Kyma Installation
type: Installation
---

This Installation guide describes steps necessary to update Kyma Installation in local and cluster environments.

## Prerequisites

- Kyma is succesfully installed, Installation CR status is `Installed`

> - Access to the Kyma installation cluster with kubectl tool

Author: I would leave this requirement. It clearly defines that only person with direct access to the cluster can perform an update.

- Docker tools to build Docker images
- Access to a Docker Registry - only for cluster installation

## Overview

Kyma consists of multiple components, installed as [Helm](https://github.com/helm/helm/tree/master/docs) releases.

> Kyma update means introducing changes to these releases - changes in charts, changes in override values or both. Adding new Helm releases is also considered an update of an existing installation.

Reviewer: ```suggestion
Update of an existing installation can include:
- changes in charts
- changes in overrides
- adding new releases
```

> Some updates might not be possible, because of breaking changes between component versions or dependency conflicts. In that case a component-specific work-around is necessary. Such component-specific custom procedures or work-arounds are out of scope of this document. Similarly, any other custom update-related steps, like data migration, are out of scope.

Reviewer: ```suggestion
> **NOTE:** In case of dependency conflicts or major changes between components versions, some updates may not be possible. Further steps assume that the update of all involved components is not obstructed.
```

> Further steps assume that update of all involved components is possible.

Reviewer: ```suggestion
```

> ## Update process overview

Reviewer: IMHO this section is redundant and can be skipped.

Kyma update process is configured by the Installation CR instance described [here](./050-cr-installation.md) and is performed by the Installer.

The Installer is activated by setting a label `action=install` on the Installation CR instance.

Upon activation the Installer loops over component list defined in the Installation CR instance. For every component on the list, the Installer:
- Reads corresponding overrides for component as described [here](./044-gs-installation-overrides.md)
- Updates the component with `helm upgrade <component>` if the component is already installed
- Installs the component with `helm install <component>` if the component is new.

> Currently there's no support for components removal during update. This has to be done manually.

Reviewer: This could be mentioned as a note in the overview section.

## Update procedure

The update procedure consists of three main steps:
- Prepare the update
- Build updated Kyma-Installer image and redeploy it
- Trigger the update process

### Prepare the update

> An update contains changes in charts, changes in override values, or both.

Reviewer: It is mentioned at the beginning of the document, no need to reiterate it here.

> Make all required changes to Helm charts in the `resources` subdirectory.

Reviewer: ```suggestion
For existing releases: Make all required changes to Helm charts in the `resources` subdirectory.
```

> Add new components to be included in the update. These are top-level Helm charts created in the `resources` directory. For any new component, add it to the Installation CR components list, as described [here](./050-cr-installation.md), otherwise the Installer will ignore it during update.

Reviewer: ```suggestion
For new components: add a top-level Helm chart for each component. Use `kubectl edit installation kyma-installation` to include the component on the Installation CR components list, as described in **Installation Custom Resource** 
```

Author: Done.

> The easiest way to add new component to the list is to directly edit Installation CR instance with: `kubectl edit installation kyma-installation`

Reviewer: ```suggestion
```

Author: Done.

> Update ConfigMaps and Secrets containing override values for the Installer, as described [here](./044-gs-installation-overrides.md). You can also add entirely new ConfigMaps and Secrets.

Author: Done.

- Access to the Kyma installation cluster with kubectl tool
- Docker
- Access to a Docker Registry - only for cluster installation

## Overview

Kyma consists of multiple components, installed as [Helm](https://github.com/helm/helm/tree/master/docs) releases.

Update of an existing installation can include:
- changes in charts
- changes in overrides
- adding new releases

> **NOTE:** In case of dependency conflicts or major changes between components versions, some updates may not be possible. Further steps assume that the update of all involved components is not obstructed.

> **NOTE:** Currently there's no support for components removal during update.

## Update procedure

The update procedure consists of three main steps:
- Prepare the update
- Update Kyma-Installer
- Trigger the update process

### Prepare the update

For existing components: Make all required changes to Helm charts in the `resources` subdirectory.

For new components: add a top-level Helm chart for each component. Use `kubectl edit installation kyma-installation` to include the component on the Installation CR components list, as described in the **Installation Custom Resource** document.

> For changes in overrides: update existing ConfigMaps and Secrets. Add new ConfigMaps and Secrets if required. See the **Installation overrides** document for further reference.m-

Author: Fixed.

### Update Kyma-Installer

#### Update Kyma-Installer - local scenario

- Build new image for Kyma-Installer with updated charts using the following script:  
```
./installation/scripts/build-kyma-installer.sh
```  
> **NOTE:** If the existing installation was started using `installation/cmd/run.sh` script with a `--vm-driver <value>` parameter, provide the same parameter to the `build-kyma-installer.sh` script.

- Redeploy the Kyma-Installer Pod:  
```
kubectl delete pod -n kyma-installer <Installer-Pod-name>
```

#### Update Kyma-Installer - cluster scenario

- Build new image for Kyma-Installer with updated charts using the following command:  
```

> docker build -t <your_image_name>:<your_tag> -f tools/kyma-installer/kyma.Dockerfile . --build-arg INSTALLER_VERSION=63484523

Author: "--build-arg [...]" removed.

---
title: Update Kyma
type: Installation
---

This guide describes how to update Kyma deployed locally or on a cluster.

## Prerequisites

- [Docker](https://www.docker.com/)
- Access to a Docker Registry - only for cluster installation

## Overview

Kyma consists of multiple components, installed as [Helm](https://github.com/helm/helm/tree/master/docs) releases.

Update of an existing deployment can include:
- changes in charts
- changes in overrides
- adding new releases

The update procedure consists of three main steps:
- Prepare the update
- Update the Kyma Installer
- Trigger the update process

> **NOTE:** In case of dependency conflicts or major changes between components versions, some updates may not be possible.

> **NOTE:** Currently Kyma doesn't support removing components as a part of the update process.

## Prepare the update

> - If you update and existing component, make all required changes to the Helm charts of the component located in the in the `[resources](https://github.com/kyma-project/kyma/tree/master/resources)` directory.

Reviewer 2: you cannot use the code ` sings like this because your sentence cannot be rendered in a proper way, see
<img width="900" alt="screen shot 2019-01-15 at 14 55 18" src="https://user-images.githubusercontent.com/17568639/51184845-c7298b00-18d5-11e9-8249-402235374f62.png">

if you want to have the **resources** in a block code then you need to move your ` sings inside the brackets, like this one:
```
[`resources`](https://github.com/kyma-project/kyma/tree/master/resources)  directory.
```

this will produce such output 
[`resources`](https://github.com/kyma-project/kyma/tree/master/resources) directory.

Author: Thanks, Fixed.

- If you add a new component to your Kyma deployment, add a top-level Helm chart for that component. Additionally, run this command to edit the Installation custom resource and add the new component to the installed components list:
  ```
  kubectl edit installation kyma-installation
  ```

>   > **NOTE:** Read [this](#custom-resource-installation) document to learn more about the Installation custom resource.

Reviewer 2: this link is not working
